src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -85,38 +86,29 @@ public StepExecution start(StepContext context) throws Exception {
         return new Execution(this, context);
     }
 
-    public static final class Execution extends StepExecution {
+    static final class Execution extends AbstractStepExecutionImpl {
 
         private static final long serialVersionUID = 1;
 
         private transient BindingStep step;
 
-        public Execution(@Nonnull BindingStep step, StepContext context) {
+        Execution(@Nonnull BindingStep step, StepContext context) {
             super(context);
             this.step = step;
         }
 
         @Override public boolean start() throws Exception {
             Run<?,?> run = getContext().get(Run.class);
-            if (run == null) {
-                throw new MissingContextVariableException(Run.class);
-            }
             TaskListener listener = getContext().get(TaskListener.class);
-            if (listener == null) {
-                throw new MissingContextVariableException(TaskListener.class);
-            }
 
             FilePath workspace = getContext().get(FilePath.class);
             Launcher launcher = getContext().get(Launcher.class);
 
             Map<String,String> overrides = new HashMap<String,String>();
             List<MultiBinding.Unbinder> unbinders = new ArrayList<MultiBinding.Unbinder>();
             for (MultiBinding<?> binding : step.bindings) {
-                for (Class<?> requiredContext : binding.getDescriptor().getRequiredContext()) {
-                    Object v = getContext().get(requiredContext);
-                    if (v == null) {
-                        throw new MissingContextVariableException(requiredContext);
-                    }
+                if (binding.getDescriptor().requiresWorkspace() && workspace == null) {
+                    throw new MissingContextVariableException(FilePath.class);
```

Comment:
Technically you should also check for `Launcher`, though in practice they will both be present or neither.